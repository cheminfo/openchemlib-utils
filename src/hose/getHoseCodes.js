import { ensureHeterotopicChiralBonds } from '../diastereotopic/ensureHeterotopicChiralBonds.js';
import { tagAtom } from '../util/tagAtom.js';

import { getHoseCodesForAtomsInternal } from './getHoseCodesForAtomsInternal.js';

/**
 * Returns the hose codes for all atoms in the molecule
 * @param {*} molecule
 * @param {object} [options={}]
 * @param {string[]} [options.atomLabels]
 * @param {number} [options.minSphereSize=0]
 * @param {number} [options.maxSphereSize=4]
 * @returns
 */
export function getHoseCodes(molecule, options = {}) {
  const { atomLabels, minSphereSize, maxSphereSize } = options;
  const { Molecule } = molecule.getOCL();

  const atomicNumbers = atomLabels?.map((label) =>
    Molecule.getAtomicNoFromLabel(label),
  );

  const internalMolecule = molecule.getCompactCopy();
  internalMolecule.addImplicitHydrogens();
  ensureHeterotopicChiralBonds(internalMolecule);

  const results = [];

  for (let i = 0; i < internalMolecule.getAllAtoms(); i++) {
    if (
      atomicNumbers &&
      !atomicNumbers.includes(internalMolecule.getAtomicNo(i))
    ) {
      results.push(undefined);
    } else {
      const tempMolecule = internalMolecule.getCompactCopy();
      tagAtom(tempMolecule, i);
      results.push(
        getHoseCodesForAtomsInternal(tempMolecule, {
          minSphereSize,
          maxSphereSize,
        }),
      );
    }
  }

  return results;
}
