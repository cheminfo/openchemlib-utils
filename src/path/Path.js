import { getPathsInfo } from './getPathsInfo';

export const Path = {
  getPathsInfo,
};
