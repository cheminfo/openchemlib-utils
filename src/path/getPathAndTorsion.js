/**
 * Calculates the path between 2 atoms
 * @param {import('openchemlib').Molecule} molecule
 * @param {number} from - index of the first atom
 * @param {number} to - index of the end atom
 * @param {number} maxLength - maximal length of the path
 */
export function getPathAndTorsion(molecule, from, to, maxLength) {
  let originalAtoms = []; // path before renumbering
  molecule.getPath(originalAtoms, from, to, maxLength + 1);
  let torsion;
  if (originalAtoms.length === 4) {
    torsion = molecule.calculateTorsion(originalAtoms);
  }

  return {
    atoms: originalAtoms,
    from,
    to,
    torsion,
    length: originalAtoms.length - 1,
  };
}
