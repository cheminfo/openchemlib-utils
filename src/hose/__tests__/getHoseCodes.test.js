import { readFileSync } from 'fs';
import { join } from 'path';

import { Molecule } from 'openchemlib';
import { expect, it, describe } from 'vitest';

import { getHoseCodes } from '../getHoseCodes.js';

describe('getHoseCodes', () => {
  it('methane', () => {
    const molecule = Molecule.fromSmiles('C');
    const hoseCodes = getHoseCodes(molecule, { maxSphereSize: 1 });
    expect(hoseCodes).toStrictEqual([
      ['fH@NJ`\x7FRapj`', 'fH@FJ`\x7FRapj`'],
      ['fHdrA\x7FRaDj`', 'eFBBYcA\x7FRPQJh'],
      ['fHdrA\x7FRaDj`', 'eFBBYcA\x7FRPQJh'],
      ['fHdrA\x7FRaDj`', 'eFBBYcA\x7FRPQJh'],
      ['fHdrA\x7FRaDj`', 'eFBBYcA\x7FRPQJh'],
    ]);
  });

  it('Hc1ccccc1', () => {
    const molecule = Molecule.fromSmiles('Cc1ccccc1');
    molecule.setAtomicNo(0, 1);
    const hoses = getHoseCodes(molecule, { maxSphereSize: 2 });
    const distincts = getDistinctHoses(hoses);
    expect(distincts.length).toBe(6);
    expect(distincts).toMatchSnapshot();
  });

  it('ethanol only C and O', () => {
    const molecule = Molecule.fromSmiles('CCO');
    const hoseCodes = getHoseCodes(molecule, {
      atomLabels: ['C', 'O'],
      maxSphereSize: 0,
    });
    expect(hoseCodes).toStrictEqual([
      ['fH@NJ`\x7FRapj`'],
      ['fH@NJ`\x7FRapj`'],
      ['fI@GEP_iP~UP'],
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
    ]);
  });

  it('cyclosporin', () => {
    const molfile = readFileSync(
      join(__dirname, '../../diastereotopic/__tests/data/cyclosporin.mol'),
      'utf8',
    );
    const molecule = Molecule.fromMolfile(molfile);

    let hoseCodes = getHoseCodes(molecule, { maxSphereSize: 1 });
    expect(hoseCodes.length).toBe(196);
    let distincts = getDistinctHoses(hoseCodes);
    expect(distincts.length).toBe(20);
    expect(distincts).toMatchSnapshot();

    hoseCodes = getHoseCodes(molecule, { maxSphereSize: 1, atomLabels: ['C'] });
    expect(hoseCodes.length).toBe(196);
    distincts = getDistinctHoses(hoseCodes);
    expect(distincts.length).toBe(10);
    expect(distincts).toMatchSnapshot();

    hoseCodes = getHoseCodes(molecule, { atomLabels: ['C'] });
    expect(hoseCodes.length).toBe(196);
    expect(hoseCodes).toMatchSnapshot();
  });
});

function getDistinctHoses(hoses) {
  const distinct = {};
  for (const hose of hoses) {
    if (!hose) continue;
    for (const sphere of hose) {
      distinct[sphere] = true;
    }
  }
  return Object.keys(distinct);
}
