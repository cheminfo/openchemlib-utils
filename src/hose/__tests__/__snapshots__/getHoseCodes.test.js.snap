// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHoseCodes Hc1ccccc1 1`] = `
[
  "fH@NJ\`Rapj\`",
  "eM@HpCbJCtQ~d\`xUP",
  "gJP@DiP@CbB\`}aDaQ~dPNET",
  "fHdrARaDj\`",
  "eFBBYcARPQJh",
  "gC\`HALiM@ARHDRj@",
]
`;

exports[`getHoseCodes cyclosporin 1`] = `
[
  "fH@NJ\`Rapj\`",
  "gCh@@duPGDeAzROtbQpj\`",
  "fJ@GEP_iPnUP",
  "gCh@AGj@x\`hOtbA\\j\`",
  "gCi@DDfZ@xdhOtbQpj\`",
  "gCa@@duPGDeAzROtbQpj\`",
  "eF\`BLGETGj{}IIpj\`",
  "eM\`BN\`xb\`}D_iHKeT",
  "fI@GEP_iP~UP",
  "eFHBJGDTGzRCyU@",
  "gC\`@H}PGDEAzBOtbApj\`",
  "eFHBLGDTGhc}IA|j\`",
  "eM@HzCbJCtY~d\`xUP",
  "eF@Hp\\QP^cotdGBj@",
  "eM@HvCbJCtQ~d\`xUP",
  "eM\`AIhNJhOUgzRSaU@",
  "fHdrARaDj\`",
  "eFBBYcARPQJh",
  "eFbHfXp_tdDRj@",
  "eFJHfXp_tdDRj@",
]
`;

exports[`getHoseCodes cyclosporin 2`] = `
[
  "fH@NJ\`Rapj\`",
  "gCh@@duPGDeAzROtbQpj\`",
  "gCi@DDfZ@xdhOtbQpj\`",
  "gCa@@duPGDeAzROtbQpj\`",
  "eF\`BLGETGj{}IIpj\`",
  "gC\`@H}PGDEAzBOtbApj\`",
  "eM@HzCbJCtY~d\`xUP",
  "eF@Hp\\QP^cotdGBj@",
  "eM@HvCbJCtQ~d\`xUP",
  "eM\`AIhNJhOUgzRSaU@",
]
`;

exports[`getHoseCodes cyclosporin 3`] = `
[
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "deeD@DHAR[fYjjiLHj\`NHr\`|fQT_iBXxUP",
    "do]L@DhIwHheMEECFjjijSBFhCbLhOLdYCPOSBq~dIcaU@",
    "fbyq\`@DTbCUlbbTTTTLLtSUUUSUTSACJCDNHAqCEAyRHqAHYBnuAh]GlCwzPSCaU@",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "gNu@EPdrsUTq@xdhOIQhRIGBj@",
    "dkMD@DhMRYYfYjjjiLHZ\`NHb\`|RQdECQ~dICaU@",
    "fbypP@DT\\CU\`yEDhhhhheCFjjjffkFBBTLhAqBEAxRHqAHY@lUvhMGzPRCaU@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCa@@duPGDeAzROtbQpj\`",
    "dazH@BAIfZjjRJDHCbHhODDQBQd_iBPxUP",
    "dkMD@DHArJZJJVIjjjkLhhbZ@xbjCsYEPLSByo_iBTxUP",
    "fhaQ\`@DTAKVQQJZJJFGMUUSUNLTLHUPCbCJCtdQbFPKiPjKX{oQ~dDXxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "dayH@DXDfYfjjLHCbHhOQd_iBPxUP",
    "dgmH@DXDfYfyfjjjjLHCbHhOEdEE]t_iBPxUP",
    "fbypP@DL|ApTiLsMsLsMUUUSTvLDDhePCbDJCplPRJ]rIQkc][ot\`dGBj@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF\`BLGETGj{}IIpj\`",
    "gCh@AGj@xdhOR]~dRNET",
    "dazH@DCHhhXuU@GDEP^PnjOtaB\\Jh",
    "dgoL@D[PT|bbbfabcUUSUfDUQM@\\PUAxl\`njIc}HPgBj@",
  ],
  undefined,
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gC\`@H}PGDEAzBOtbApj\`",
    "gNq@@ds]URaBhGDUAxrEAr}~dQNET",
    "defH@FCHhhhkUUSdDIQa@\\PeA{\`bHRL\`lkac}HQGBj@",
    "do]D@DHArJZJJVKIjjjilrbbIh\`{bIhOCdU@qLKf|\\WGzPdnET",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "deeH@DHDfyfZjjcB@xcJCrYEQ~dIcaU@",
    "dgmD@DXCrJJJFJJMUUTtXPGDYP^kHzIy^wotaL\\Jh",
    "fdqp\`@DLBAvQQQQSQPqqjjjjZdXH|PCbFJCplPRJ]rM\\[k]zOt\`fGBj@",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "dayH@DHDfyVjjLHCbLhOYdEEY~dIcaU@",
    "dcuD@DXErJJJEJMUUMFDAqFTGbrFclgotaL\\Jh",
    "do_D@DZpwHhhhlmCFjjijSBCHCbLhO]dEe\\t]b|\\_iBXxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eM@HzCbJCtY~d\`xUP",
    "gGP@DiZf\`NHJCvFREGzQ@xUP",
    "diDH@@RYoZjfRJDHakbBhOMDIAYWMGQ~dHSaU@",
    "dmVH@FCHhhhjjjj[HhPcJHFx\`jCtaDPdYAYWCDpo_iBDxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gNq@AdTTMURaBHGDMAxrEAzy~dPnET",
    "defH@FCHhhicUUSdTIQe@\\PUA{\`bHRL\`nkAc}HPgBj@",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "dayH@DXDfYfjjLHCbHhOQd_iBPxUP",
    "dmUH@DXDfYgVjjjLHCbHhOUdEe\\t]gzPdNET",
    "do_L@DXpvtfYgVYjjjYhp\`NHb\`|vPVUsQvKwzPdNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eM@HzCbJCtY~d\`xUP",
    "gJX@@dsURTCbJ\`}iBp}~dQNET",
    "dayH@DHDfyVjjLhCbJhOEdEEYw_iBTxUP",
    "dcuD@DXErJJJEJMUUMFTAqGTGbrFclgotaN\\Jh",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eM@HvCbJCtQ~d\`xUP",
    "gJP@DiVhPWDEA{BIBs}H\`\\Jh",
    "gNp@Di^ZjDEqAP^hbPlwFOtbApj\`",
    "diDH@@RYoZjfRBDhakbFhO]DIAYWMGPO_iBLxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "dayH@DXDfYfjjLHCbHhOQd_iBPxUP",
    "dkMH@DXDfYfyZjjjcB@xbJCqYAQW]gzPdNET",
    "fjeq\`@DL\\BtTfYfyYfyjjjijZdXHIPCbDJCwLPRJ]sIQkcQ~dD\`xUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF\`BLGETGj{}IIpj\`",
    "gCh@AGj@xdhOR]~dRNET",
    "dazH@DCHhhXuU@GDEP^PnjOtaB\\Jh",
    "dgoD@D[PWHhhihXhuUTuIaETAqATGbrBzhfOtaB\\Jh",
  ],
  undefined,
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "eM@Df\`xb\`|gV_iHNET",
    "gJX@@dsURTCbF\`}iBp}~dPnET",
    "dayH@DHDfyVjjLHCbNhOEdEEYw_iB\\xUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eM@HvCbJCtQ~d\`xUP",
    "gC\`@DiZDEqAP^pbPnRHGBj@",
    "gJP@DivhPWDEAxbIB{Y~dPNET",
    "gNp@DivZjDEqAP^hbPnvFOtbApj\`",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "diEH@DHDfyejjhp\`NHr\`|fQV_iBXxUP",
    "do]D@DXMrJJJFISQjjjZhp\`NHr\`|VQvSrq~dIcaU@",
    "flip\`@DLbBvQQQQSQFZZMUUUMUHpPx\`GDLTGiX\`dT{fZxtNcwA{}HIapj\`",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eM\`AIhNJhOUgzRSaU@",
    "gNu@DPdvsUPGDuAz[OtbYpj\`",
    "dmUD@DXIrJJHqQjjf@NHr\`|fXO_iBXxUP",
    "do_D@DZpwHhheMCCFjjij\`r@xcJCuYaQWKqq~dIcaU@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF\`BLGETGj{}IIpj\`",
    "gCh@AGj@xdhOR]~dRNET",
    "gNy@DLbbaZj@xahOH}\\RHWBj@",
    "dmWD@DYPgHhhicFjjZLHCbBhOYdEuY~dHSaU@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "eM@DfPxb\`|gT_iHNET",
    "gC\`@DefDeqAP^pzPhRHGBj@",
    "gJP@DeYhRWDEAxciBcY~dPNET",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eM@HzCbJCtY~d\`xUP",
    "gNx@@dsuUTe@xbhOZPlL_iDSaU@",
    "dmUH@DHLbfbbLZjjhr\`NHj\`}VQVCD|o_iBTxUP",
    "do]D@DXMrJJJFISQjjjZhr\`NHz\`|vQvSrq_OwzPgNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "gNu@DpdsKUPGDeAz[OtbQpj\`",
    "dmUH@DXDfY[fjjh@xbJCvYaQW_iBPxUP",
    "f\`~q\`@DLLCSdfY[fVZjjff\`R\`GDHTGjXpdTxVJ{}HIApj\`",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "dayH@DXDfYfjjLHCbHhOQd_iBPxUP",
    "dkMH@DXDfYfyZjjjcB@xbJCqYAQW]gzPdNET",
    "fbyq\`@DL\\BpTfYfyY[fjjjffiFBBT@xaB\`}sDDbg\\rVZxt_iAHNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF\`BLGETGj{}IIpj\`",
    "gCh@AGj@xdhOR]~dRNET",
    "dazH@DCHhhXuU@GDEP^PnjOtaB\\Jh",
    "dkOD@D[PWHhhihXfjjfdpbj@x\`jCqYA]TSgzPaNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gC\`@H}PGDEAzBOtbApj\`",
    "gJP@DjZhCbB\`|QDq]nRHGBj@",
    "gNx@@dsuUTe@xahOVPlLWgwzQBxUP",
    "dmUH@DHLbfbbLZjjhp\`NHF\`}VQVCD|o_iBBxUP",
  ],
  undefined,
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "diEH@DHDfyejjhp\`NHr\`|fQV_iBXxUP",
    "do]D@DXMrJJJFISQjjjZhp\`NHr\`|VQvSrq~dIcaU@",
    "flip\`@DLbBvQQQQSQFZZMUUUMUHpPz\`GDLTGiX\`dT{fZxtNcwA{}HIapj\`",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gJP@DfZjR\`\\PTGbNdKMwzQ@xUP",
    "gNx@@dsuURaAHGDmAzrEab|~RIWBj@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gJP@DfZjR\`\\PTGbNdKMwzQ@xUP",
    "gNx@@dsuUReAHGDmAzrEab|~RIWBj@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "diEH@DHDfyejjhp\`NHr\`|fQV_iBXxUP",
    "dgmD@DhERY[enZjjZhp\`NHr\`}RQdM\`}L_iBXxUP",
    "fhap\`@DTB@vQQJJJHsSMUUTuTLD@xab\`}IDX\`dLqWZbvN{w_iALNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF\`BLGETGj{}IIpj\`",
    "gCh@AGj@xdhOR]~dRNET",
    "dazH@DCHhhXuU@GDEP^PnjOtaB\\Jh",
    "dcwD@DKPgHihhmBuUSTXPGDMP^KHkFy\\RDYpj\`",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eM@HzCbJCtY~d\`xUP",
    "gNx@@dsuUTe@xbhOZPlL_iDSaU@",
    "dmUH@DHLbfbbLZjjhr\`NHj\`}VQVCD|o_iBTxUP",
    "do]D@DXMrJJJFISQjjjZhr\`NHz\`|vQvSrq_OwzPgNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "gNu@EPdrsUTq@xdhOIQhRIGBj@",
    "dcuH@DhDfVYejjjhp\`NHb\`|RQdECY~dICaU@",
    "fbyq\`@DTl@pTfVYef[fjjjZfiFBBT@xaB\`}qDX\`dLpTJ{T_iAHNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eM@HzCbJCtY~d\`xUP",
    "gNx@@dsuUTe@xbhOZPlL_iDSaU@",
    "dmUH@DHLbfbbLZjjhr\`NHj\`}VQVCD|o_iBTxUP",
    "dgmD@DhERY[enZjjZhr\`NHZ\`}rQdM\`}LGu}~dHsaU@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "dayH@DXDfYfjjLHCbHhOQd_iBPxUP",
    "dmUH@DXDfYfvjjjLHCbHhOUdEE\\t]wzPdNET",
    "f\`~q\`@DLLCSdfYf~VZjjffQb\`e@NHPhOLqAHisHytlRBP\\Jh",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF\`BLGETGj{}IIpj\`",
    "gCh@AGj@xdhOR]~dRNET",
    "dazH@DCHhhXuU@GDEP^PnjOtaB\\Jh",
    "dkOD@D[PWHhhihXfjjfdrbj@x\`jCqYA]TSgzPaNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gC\`@H}PGDEAzBOtbApj\`",
    "gJP@DjZhCbB\`|QDq]nRHGBj@",
    "gNx@@dsuUTe@xahOVPlLWgwzQBxUP",
    "dmUH@DHLbfbbLZjjhp\`NHF\`}VQVCD|o_iBBxUP",
  ],
  undefined,
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gC\`@H}PGDEAzBOtbApj\`",
    "gJP@DjZhCbB\`|QDq]nRHGBj@",
    "gNx@@dsuUTe@xahOVPlLWgwzQBxUP",
    "dmUH@DHLbfbbLZjjhp\`NHF\`}VQVCD|o_iBBxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "gNu@DPdwKUTq@xfhO[PXo_iDsaU@",
    "dmUD@DXIrJJJGQjjfcB@xcJCvYCQw_iBXxUP",
    "do_D@DZpwHhhhikMFjjijSBChCbLhOMdEE\\t]qq~dIcaU@",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gJP@DfZjR\`\\PTGbNdKMwzQ@xUP",
    "gNx@@dsuURaAHGDmAzrEab|~RIWBj@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gJP@DfZjR\`\\PTGbNdKMwzQ@xUP",
    "gNx@@dsuUReAHGDmAzrEab|~RIWBj@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "diEH@DHDfyYjjhp\`NHr\`}fPTUGzPfNET",
    "dkMD@DXMrJJJEQQjjjZLHCbLhOUdMGQOKwzPfNET",
    "fhap\`@DL|BvQQQQSJFVMUUUMRLDNHAqCEAxVHIENYGHuqn]GzPSCaU@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gJP@DfZjR\`\\PTGbNdKMwzQ@xUP",
    "gNx@@dsuURaAHGDmAzrEab|~RIWBj@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gJP@DfZjR\`\\PTGbNdKMwzQ@xUP",
    "gNx@@dsuUReAHGDmAzrEab|~RIWBj@",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gCh@@duTe@xbhOJPnRHgBj@",
    "gNu@DPdwKUTu@xehO[PXo_iDkaU@",
    "dmUD@DXIrJJJGQjjfcJ@xcjCvYCQw_iB\\xUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "dayH@DXDfYfjjLHCbHhOQd_iBPxUP",
    "dkMH@DXDfYfvZjjjcB@xbJCuYAQWMGQ~dICaU@",
    "fdqq\`@DL\\BpTfYfvYYjjjiidXHIPCbDJCplPRJ\\rNQKM\\[ot\`dGBj@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gC\`@H}PGDEAzBOtbApj\`",
    "gGX@@dsUUIPNHjCqdJCu{}Hb\\Jh",
    "diEH@DHDfyYjjhr\`NHj\`}VPTUG\\O_iBTxUP",
    "dkMD@DXMrJJJEQQjjjZLhCbNhOUdMGQOKwzPgNET",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "gNu@EPdrsUTq@xdhOIQhRIGBj@",
    "deeH@DhDfVYZjjcB@xbJCuIFPTUC]~dICaU@",
    "dgoL@DkPTtfVY{fjjfYLHH\`NHb\`|rQdEEPwCwzPdNET",
  ],
  undefined,
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gGX@@dsUThPbAqCP^LaP^o_iDKaU@",
    "diEH@DHDfyYjjdpbb@xcjCuYAQT]p}~dIsaU@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gGX@@dsUTiPbAqCP^LaP^o_iDKaU@",
    "diEH@DHDfyYjjdpbj@xcjCuYAQT]p}~dIsaU@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "gNu@DPdwKUTq@xfhO[PXo_iDsaU@",
    "deeD@DhARY[yjjZLHCbLhODdYEPw_iBXxUP",
    "dgoD@DhpWHheEEYhuUTuIaFdAqFTGfRL\`hjF{XRDqpj\`",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gCh@@duPGDeAzROtbQpj\`",
    "diEH@DHDfyejjhp\`NHr\`|fQV_iBXxUP",
    "dgmD@DhERY[enZjjZhp\`NHr\`}RQdM\`}L_iBXxUP",
    "fdqp\`@DTB@vQQJJJHsSQjjjfjdXHlPCbFJCtdQbBPsE]jKP{o]~dDpxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF\`BLGETGj{}IIpj\`",
    "gCh@AGj@xdhOR]~dRNET",
    "dazH@DCHhhXuU@GDEP^PnjOtaB\\Jh",
    "dkOD@D[PWHhhihXfjjfdpbj@x\`jCqYA]TSgzPaNET",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gCh@@duTe@xbhOJPnRHgBj@",
    "gNu@DPdwKUTu@xehO[PXo_iDkaU@",
    "deeD@DhARY[yjjZLhCbFhODdYEPw_iBLxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gCi@DDfZ@xdhOtbQpj\`",
    "gNu@EPdrsUTq@xdhOIQhRIGBj@",
    "dcuH@DhDfVYejjjhp\`NHb\`|RQdECY~dICaU@",
    "fdqq\`@DTl@sdfVYenyjjjfZdXHIPCbDJCwDQbBPsA\\kmQ~dD\`xUP",
  ],
  undefined,
  [
    "fH@NJ\`Rapj\`",
    "gC\`@H}PGDEAzBOtbApj\`",
    "gGX@@dsUUIPNHjCqdJCu{}Hb\\Jh",
    "deeH@DHDfyfZjjcJ@xbjCqYEPOSwzPeNET",
    "dgmD@DXCrJJJFJJMUUTtYPGD]P^kHzIy^wotaN\\Jh",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gGX@@dsUThPbAqCP^LaP^o_iDKaU@",
    "deeH@DHDfyfZjjSBJHCbAhOEdU@}O_iBBxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gGX@@dsUTiPbAqCP^LaP^o_iDKaU@",
    "deeH@DHDfyfZjjSBJhCbAhOEdU@}O_iBBxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eM@HzCbJCtY~d\`xUP",
    "gNx@@dsuUTe@xbhOZPlL_iDSaU@",
    "dmUH@DHLbfbbLZjjhr\`NHj\`}VQVCD|o_iBTxUP",
    "dgmD@DhERY[enZjjZhr\`NHZ\`}rQdM\`}LGu}~dHsaU@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "gC\`@H}PGDEAzBOtbApj\`",
    "gJP@DjZhCbB\`|QDq]nRHGBj@",
    "gNx@@dsuUTe@xahOVPlLWgwzQBxUP",
    "dmUH@DHLbfbbLZjjhp\`NHF\`}VQVCD|o_iBBxUP",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gJP@DfZjR\`\\PTGbNdKMwzQ@xUP",
    "gNx@@dsuURaAHGDmAzrEab|~RIWBj@",
  ],
  [
    "fH@NJ\`Rapj\`",
    "eF@Hp\\QP^cotdGBj@",
    "gC\`@Dfj@x\`hOH]HRHGBj@",
    "gJP@DfZjR\`\\PTGbNdKMwzQ@xUP",
    "gNx@@dsuUReAHGDmAzrEab|~RIWBj@",
  ],
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
]
`;
